File: domain/src/main/java/com/study/platform/domain/apply/model/ApplyRepository.java
package com.study.platform.domain.apply.model;

import java.util.Optional;
import java.util.UUID;

public interface ApplyRepository {

    Apply save(Apply apply);
    void delete(Apply apply);
    Optional<Apply> findById(UUID id);

    boolean existsByPostIdAndApplicantId(UUID postId, UUID applicantId);
    Optional<Apply> findByIdWithPostAndApplicant(UUID applyId);
    Optional<Apply> findByPostIdAndApplicantId(UUID postId, UUID applicantId);
    long countByPostIdAndStatus(UUID postId, ApplyStatus status);
    Optional<Apply> findByIdWithPostAndApplicantForUpdate(UUID applyId);
}

File: domain/src/main/kotlin/com/study/platform/domain/apply/model/ApplyRepository.kt
package com.study.platform.domain.apply.model

import java.util.UUID

interface ApplyRepository {

    fun save(apply: Apply): Apply
    fun delete(apply: Apply)
    fun deleteAllByPostId(postId: UUID)
    fun findById(id: UUID): Apply?

    fun existsByPostIdAndApplicantId(postId: UUID, applicantId: UUID): Boolean
    fun findByPostIdAndApplicantId(postId: UUID, applicantId: UUID): Apply?
    fun countByPostIdAndStatus(postId: UUID, status: ApplyStatus): Long
    fun countByPostIdAndStatusForUpdate(postId: UUID, status: ApplyStatus): Long
    fun findByIdWithPostAndApplicantForUpdate(applyId: UUID): Apply?
    fun findByIdForUpdate(applyId: UUID): Apply?
    fun findPostIdByApplyId(applyId: UUID): UUID?
}
